framework/src/headers/ip/v6/icmp/ndp_options.rs

```diff
@@ -0,0 +1,228 @@
+        NdpOptionType::SourceLinkLayerAddress
+    }
+}
+
+/// Provides functions to implementors that allow the implementor to scan their NDP message for
+/// NDP options.  The type of options available are specific to the type of NDP Message
+/// Ensures that implementors support NdpMessageContents, which further enforces EndOffset + Sized
+/// which we require to do parsing
+pub trait NdpOptions: NdpMessageContents {
+    /// Attempts to lookup the source link layer address from the underlying NDP Message
+    fn get_source_link_layer_address_option(&self, payload_length: u16) -> Option<&MacAddress> {
+        let opt = self.option_scan::<LinkLayerAddressOption>(payload_length);
+        opt.map(|sll| &sll.addr)
+    }
+
+    /// Attempts to lookup the mtu option from the underlying NDP Message
+    fn get_mtu_option(&self, payload_length: u16) -> Option<u32> {
+        let mtu_option = self
+            .option_scan::<MtuOption>(payload_length)
+            .map(|mtu| mtu.get_mtu());
+        mtu_option
+    }
+
+    /// Searches through a packet for an option, returns the first it finds or none
+    /// Relies on the caller to specify T, and they must know that the type can be parsed...
+    /// this is unfortunate, but I do not have a compile time way as the NDPOptionType is something
+    /// like the number 1, and that has to map to a MacAddress, like how do you enforce that?
+    fn option_scan<T: NdpOption>(&self, payload_length: u16) -> Option<&T> {
+        unsafe {
+            // the start of the options we offset from is the start of the ndp message
+            let opt_start = (self as *const Self) as *const u8;
+            let mut payload_offset = self.offset();
+
+            // make sure we do not scan beyond the end of the packet
+            while (payload_offset as u16) < payload_length {
+                // we want to pull the first two bytes off, option type and option length
+                let seek_to = opt_start.add(payload_offset);
+                let option_meta = slice::from_raw_parts(seek_to, 2);
+
+                // first option meta is the option type, let's cast it to our NDPOptionType
+                let cur_type: NdpOptionType = FromPrimitive::from_u8(option_meta[0]).unwrap();
+
+                // second option is the length in octets (8 byte chunks).  So a length of 1 means 8 bytes
+                let cur_size = option_meta[1] * 8;
+
+                if cur_size == 0 {
+                    // Don't know how we will be here, but it is an error condition as we will
+                    // go into an infinite loop and blow the stack if we skip, so exit None
+                    return None;
+                } else if (payload_offset + cur_size as usize) > (payload_length as usize) {
+                    // we need to protect ourselves in case the option_length goes past the payload length
+                    return None;
+                }
+
+                let requested_type = T::get_type();
```

Comment:
move this variable to the top so it only happens once (i.e. before entering the while loop)